servant/src/Servant/API/UVerb.hs

```diff
@@ -0,0 +1,123 @@
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+-- | An alternative to 'Verb' for end-points that respond with a resource value of any of an
+-- open union of types, and specific status codes for each type in this union.  (`UVerb` is
+-- short for `UnionVerb`)
+--
+-- This can be used for returning (rather than throwing) exceptions in a server as in, say
+-- @'[Report, WaiError]@; or responding with either a 303 forward with a location header, or
+-- 201 created with a different body type, depending on the circumstances.  (All of this can
+-- be done with vanilla servant-server by throwing exceptions, but it can't be represented in
+-- the API types without something like `UVerb`.)
+--
+-- See <https://docs.servant.dev/en/stable/cookbook/uverb/UVerb.html> for a working example.
+module Servant.API.UVerb
+  ( UVerb,
+    Union,
+    HasStatus (StatusOf),
+    statusOf,
+    HasStatuses (Statuses, statuses),
+    WithStatus (..),
+    module Servant.API.UVerb.OpenUnion,
+    collapseUResp,
+    extractUResp,
+  )
+where
+
+import Control.Monad.Identity
+import Data.Aeson (FromJSON, ToJSON)
+import Data.Functor.Identity (Identity)
+import Data.SOP.BasicFunctors (K (K))
+import Data.SOP.Constraint (Constraint, All)
+import Data.SOP.NS (NS, cmap_NS, collapse_NS)
+import Data.Typeable (Typeable, Proxy (Proxy), cast)
+import qualified GHC.Generics as GHC
+import GHC.TypeLits (Nat)
+import Network.HTTP.Types (Status, StdMethod)
+import Servant.API.ContentTypes (MimeRender (mimeRender), MimeUnrender (mimeUnrender), NoContent)
+import Servant.API.Status (KnownStatus, statusVal)
+import Servant.API.UVerb.OpenUnion
+
+class KnownStatus (StatusOf a) => HasStatus (a :: *) where
+  type StatusOf (a :: *) :: Nat
+
+statusOf :: forall a proxy. HasStatus a => proxy a -> Status
+statusOf = const (statusVal (Proxy :: Proxy (StatusOf a)))
+
+instance KnownStatus n => HasStatus (WithStatus n a) where
+  type StatusOf (WithStatus n a) = n
+
+-- | If an API can respond with 'NoContent' we assume that this will happen
+-- with the status code 204 No Content. If this needs to be overridden,
+-- 'WithStatus' can be used.
+instance HasStatus NoContent where
+  type StatusOf NoContent = 204
+
+class HasStatuses (as :: [*]) where
+  type Statuses (as :: [*]) :: [Nat]
+  statuses :: Proxy as -> [Status]
+
+instance HasStatuses '[] where
+  type Statuses '[] = '[]
+  statuses _ = []
+
+instance (HasStatus a, HasStatuses as) => HasStatuses (a ': as) where
+  type Statuses (a ': as) = StatusOf a ': Statuses as
+  statuses _ = statusOf (Proxy :: Proxy a) : statuses (Proxy :: Proxy as)
+
+newtype WithStatus (k :: Nat) a = WithStatus a
+  deriving (Eq, Show, GHC.Generic)
+
+instance (GHC.Generic (WithStatus n a), ToJSON a) => ToJSON (WithStatus n a)
+
+instance (GHC.Generic (WithStatus n a), FromJSON a) => FromJSON (WithStatus n a)
+
+instance MimeRender ctype a => MimeRender ctype (WithStatus _status a) where
```

Comment:
Are we able to derive this instance (and other related instances) with GeneralizedNewtypeDeriving?

Reply:
Or with DerivingStrategies, good idea.  I tried both, but:

> ‘MimeRender’ is not a unary constraint, as expected by a deriving clause

Reply:
You can get away with this:

```haskell
deriving anyclass instance (GHC.Generic (WithStatus n a), ToJSON a) => ToJSON (WithStatus n a)
deriving anyclass instance (GHC.Generic (WithStatus n a), FromJSON a) => FromJSON (WithStatus n a)
deriving newtype instance MimeRender ctype a => MimeRender ctype (WithStatus _status a)
deriving newtype instance MimeUnrender ctype a => MimeUnrender ctype (WithStatus _status a)
```

Requires `GeneralizedNewtypeDeriving`, `DerivingStrategies` and `DeriveAnyClass`.

Reply:
Oh, and `StandaloneDeriving`.

Reply:
BTW, actually json stuff is simplified to this:

```haskell
newtype WithStatus (k :: Nat) a = WithStatus a
  deriving (Eq, Show, GHC.Generic)
  deriving newtype (ToJSON, FromJSON)
```

And you don't even need `Generic` constraint here, just rely on GND explicitly using `DerivingStrategies`.

Reply:
Also also:

```haskell
newtype WithStatus (k :: Nat) a = WithStatus a
  deriving (Eq, Show, GHC.Generic)
  deriving newtype (ToJSON, FromJSON, MimeRender ctype, MimeUnrender ctype)
```

GND solves everything here.

Reply:
For the sake of getting this out the door, I would prefer to not optimize this any further.  It looks tempting, but at least if we start adding DerivingStrategies to the core libraries, we'll have to drop more old releases from support.

Does this make sense?  If you give me any commits that do work with ghc 8.0, I'll probably cherry-pick them.